quantum/sequencer/tests/sequencer_tests.cpp

```diff
@@ -0,0 +1,533 @@
+    }
+
+    matrix_scan_sequencer();
+    advance_time(SEQUENCER_PHASE_RELEASE_TIMEOUT);
+
+    // On step 1, none of the notes should be released
+    for (int i = SEQUENCER_TRACKS - 1; i >= 0; i--) {
+        matrix_scan_sequencer();
+        EXPECT_EQ(last_noteon, 0);
+        EXPECT_EQ(last_noteoff, 0);
+
+        advance_time_and_reset(SEQUENCER_TRACK_THROTTLE);
+    }
+
+    // A 16th lasts 125ms at tempo=120
+    matrix_scan_sequencer();
+    advance_time_and_reset(125);
+
+    /**
+     * Step 2
+     */
+
+    // The C should play as soon as we start the sequence
+    matrix_scan_sequencer();
+    EXPECT_EQ(last_noteon, MI_C);
+    EXPECT_EQ(last_noteoff, 0);
+
+    advance_time_and_reset(SEQUENCER_TRACK_THROTTLE);
+
+    // The D should play shortly after
+    matrix_scan_sequencer();
+    EXPECT_EQ(last_noteon, MI_D);
+    EXPECT_EQ(last_noteoff, 0);
+
+    // The other notes should not be played
+    for (int i = 2; i < SEQUENCER_TRACKS; i++) {
+        advance_time_and_reset(SEQUENCER_TRACK_THROTTLE);
+
+        matrix_scan_sequencer();
+        EXPECT_EQ(last_noteon, 0);
+        EXPECT_EQ(last_noteoff, 0);
+    }
+
+    advance_time(SEQUENCER_PHASE_RELEASE_TIMEOUT);
+
+    // Release happens in the reverse order
+    // No notes but the first should be released
+    for (int i = SEQUENCER_TRACKS - 1; i > 1; i--) {
+        matrix_scan_sequencer();
+        EXPECT_EQ(last_noteon, 0);
+        EXPECT_EQ(last_noteoff, 0);
+
+        advance_time_and_reset(SEQUENCER_TRACK_THROTTLE);
+    }
+
+    // The D should be released first
+    matrix_scan_sequencer();
+    EXPECT_EQ(last_noteon, 0);
+    EXPECT_EQ(last_noteoff, MI_D);
+
+    advance_time_and_reset(SEQUENCER_TRACK_THROTTLE);
+
+    // The C should be released after
+    matrix_scan_sequencer();
+    EXPECT_EQ(last_noteon, 0);
+    EXPECT_EQ(last_noteoff, MI_C);
+
+    // For steps 3 -> SEQUENCER_STEPS - 1, nothing should be played or released
+    for (int i = 3; i < SEQUENCER_STEPS; i++) {
+        // A 16th lasts 125ms at tempo=120
+        matrix_scan_sequencer();
+        advance_time_and_reset(125);
+
+        for (int j = 0; j < SEQUENCER_TRACKS; j++) {
+            matrix_scan_sequencer();
+            EXPECT_EQ(last_noteon, 0);
+            EXPECT_EQ(last_noteoff, 0);
+
+            advance_time_and_reset(SEQUENCER_TRACK_THROTTLE);
+        }
+
+        matrix_scan_sequencer();
+        advance_time(SEQUENCER_PHASE_RELEASE_TIMEOUT);
+
+        for (int j = SEQUENCER_TRACKS - 1; j >= 0; j--) {
+            matrix_scan_sequencer();
+            EXPECT_EQ(last_noteon, 0);
+            EXPECT_EQ(last_noteoff, 0);
+
+            advance_time_and_reset(SEQUENCER_TRACK_THROTTLE);
+        }
+    }
+
+    // A 16th lasts 125ms at tempo=120
+    matrix_scan_sequencer();
+    advance_time_and_reset(125);
+
+    // We should have looped back to step 0
+    matrix_scan_sequencer();
+    EXPECT_EQ(last_noteon, MI_C);
+    EXPECT_EQ(last_noteoff, 0);
```

Comment:
This test could be broken into smaller unit tests at the cost of exposing more internal state publicly. Any thoughts on that?

Reply:
This huge test has now been broken into smaller unit tests.